### PR TITLE
New architecture

### PR DESCRIPTION
**Pull request by abeersaqib**
### Working 

- Build process using esbuild, rollup & vite.
- Serving of compiled assets.
- Realtime compile server and transform server for Metaverse-Modules and loading.
- Everything else :)

### Not Working

- everything is working w.r.t break. Architecture breaks nothing.

#### `index.mjs`

- The file implements additional functions to support the production mode. It uses the `-p` flag to know whether it's in production mode or development mode. if the file is in development mode it will pass the request to the vite ssr server and will leverage the dynamic compile. 
- If the server is in production it will serve the assets under the `dist` and dynamic assets are compile through the function `dynamicImport` 
- `dynnamicImport` helper function simply calls the `metaverserfile`'s rollup.js and leverage it's `resolveId` and `load` to serve the module. 

#### `Rollup.JS`

- Whenever a module is dynamically loaded it's inside an environment where it don't the exact paths of the `Metaversefile` & `Three` so a new Babel Transformation is added for all of the metaverse modules that changes the path of the imports dynamically. 

#### `ModuleRewrite` 

- Babel transform passes the path of the moduleRewrite file which process the import path and change them with the server path dynamically. 
- Paths are matched so there won't be duplicate instances of metaversefile or ThreeJS

#### `Vite.config.js`

-  Added Build Configuration
-  Added Development Configuration
-  Create a vite plugin to support the smooth development process.

#### `Workers [FOLDER]`

- Moved workers to seperate folder and made that folder available for dynamic loading

edit by @0reo 
closes https://github.com/webaverse/app/issues/2239
closes https://github.com/webaverse/app/issues/2337
closes https://github.com/webaverse/app/issues/1935

### CODE REVIEW

**Comment by abeersaqib** [comment]
https://github.com/webaverse/totum/pull/70

**Comment by avaer** [comment]
Re: workers, what about all the microphone handlers scripts?

https://github.com/webaverse/app/blob/84f1b5e8d3604d55dc1a6d4892305f8b587e5378/avatars/avatars.js#L8

**Comment by avaer** [comment]
Does production differ from dev? How?

File: index.prod.mjs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+          return res.end('Failed to load');
+        }
+
+        id = id.replace('/@proxy/', '');
+
+        totum.load(id).then(({code, map}) => {
+          res.writeHead(200, {'Content-Type': 'application/javascript'});
+          res.end(code);
+        }).catch(e => {
+          console.warn(e);
```
Comment (avaer) [inline index.prod.mjs:67]: Catching error but not responding?

File: index.prod.mjs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        });
+      });
+    } else {
+      req.originalUrl = loadUrl;
+      return /^\/(?:@proxy)\//.test(req.originalUrl) ? proxyReq(loadUrl, res) : res.redirect(req.originalUrl);
+    }
+  } catch (e) {
+    console.warn(e);
```
Comment (avaer) [inline index.prod.mjs:75]: Not responding to errors?

File: index.prod.mjs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      totum.resolveId(loadUrl, reqURL.href).then(id => {
+        /** ID might have /@proxy/ in the start that needs to be trimmed */
+        if (!id) {
+          res.status(500);
+          return res.end('Failed to load');
+        }
+
+        id = id.replace('/@proxy/', '');
+
+        totum.load(id).then(({code, map}) => {
```
Comment (avaer) [inline index.prod.mjs:63]: Why isn't this async? There is a try catch block but then there is this async part with a different flow.

File: server/boot.mjs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    base: './',
+  });
+
+  console.log('Copying Scenes');
+  /** Copy scene directory */
+  execSync('cp -r ./scenes ./dist/scenes');
+
+  console.log('Copying Metaverse_Modules');
+  /** Copy Metaverse directory */
+  execSync('cp -r ./metaverse_modules ./dist/metaverse_modules');
```
Comment (avaer) [inline server/boot.mjs:21]: What if FS is not writable?

Also executing shell is not the best idea. Will this even work on windows?

File: vite.config.js [hunk 5]
```diff
@@ -1,28 +1,258 @@
...
+let plugins = [
+  // reactRefresh()
+];
+
+const baseDirectory = 'assets';
+const proxyModuleBase = '-proxy';
+
+let entryPoints = [
+  {
+    path: './packages/totum/index.js',
```
Comment (avaer) [inline vite.config.js:20]: These should not be hardcoded.

File: vite.config.js [hunk 6]
```diff
@@ -1,28 +1,258 @@
...
+    manifest: true,
+    rollupOptions: {
+      preserveEntrySignatures: 'strict',
+      output: {
+        // __vite_skip_esbuild__: true,
+        exports: 'named',
+        minifyInternalExports: false,
+        format: 'es',
+        manualChunks: id => {
+          if (id.includes('three/build')) {
```
Comment (avaer) [inline vite.config.js:192]: These should not be config exceptions.

File: vite.config.js [hunk 7]
```diff
@@ -1,28 +1,258 @@
...
     fs: {
       strict: true,
     },
   },
-})
+};
+
+const config = {
+  ...defaultConfig,
+  ...process.env.NODE_ENV === 'production' ? viteConfiProduction : {},
```
Comment (avaer) [inline vite.config.js:251]: > `viteConfiProduction`

typo

File: index.prod.mjs (hunk 1 above)
Comment (abeersaqib) [reply index.prod.mjs]: Resolved the error; Re-worked code

File: index.prod.mjs (hunk 2 above)
Comment (abeersaqib) [reply index.prod.mjs]: Responding to errors now

File: index.prod.mjs (hunk 3 above)
Comment (abeersaqib) [reply index.prod.mjs]: Resolved promise/async usage.

File: vite.config.js (hunk 5 above)
Comment (abeersaqib) [reply vite.config.js]: All entry points are now dynamic and read from the dependencies of package.json this means we should keep only those dependencies which can be exposed to the browsers and later on used inside the metaverse-module

File: vite.config.js (hunk 6 above)
Comment (abeersaqib) [reply vite.config.js]: manual chunking is removed. Only kept for metaverse-modules.

File: vite.config.js (hunk 7 above)
Comment (abeersaqib) [reply vite.config.js]: Resolved.

**Comment by abeersaqib** [comment]
All issues raised by @avaer are resolved. Have to revert one of the commits related to workers path and test the new dynamic path/entry points with workers and then PR will be ready for review again.

**Comment by abeersaqib** [comment]
closes 
https://github.com/webaverse/app/issues/2239
https://github.com/webaverse/app/issues/2337
https://github.com/webaverse/app/issues/1935

**Comment by abeersaqib** [comment]
Closing in favour of 
https://github.com/webaverse/app/pull/2668

**Pull request closed without merging**